### ISSUE TITLE
How to fix tab size in Mozilla browsers

### ISSUE THREAD

**Issue by RealRaven2000** [Alhadis/Unfuck-Tabstops#2]
I was desperately searching for some method to make my GitHub code (for Thunderbird Add-ons) readable again and found this repo. I require a tab size of 2 because that's what I use in all my source code - and a lot of it is "mixed" so the indentation looks useless with other tab sizes.

Anyway, although I didn't find a solution for myself here, I came up with a very primitive, Mozilla-specific quickfix, which works for me on Waterfox (I switched to WF because Firefox stopped supporting my own legacy add-ons and I haven't got time to convert everything to web extensions). 

The solution is just a single rule in [userContent.css](https://www-archive.mozilla.org/unix/customizing.html#userContent):

```css
/* adjust tab size in github */
.tab-size[data-tab-size="8"] {
	-moz-tab-size: 2 !important;
}
```

I bet this rule could be turned into something equally useful on chrome with a GreaseMonkey script or whatever the hip people are using nowadays to tweak web content.

**Comment by Alhadis** [comment]
Does the current script not work when run as a bookmarklet in Firefox? 🤔

> I was desperately searching for some method to make my GitHub code (for Thunderbird Add-ons) readable again

To change the tab-size GitHub uses when displaying source code, there's an unintuitive (and undocumented) hack you can try that involves adding an [`.editorconfig` file](https://editorconfig.org/) with an `indent_size` field. [For example](https://github.com/Alhadis/Utils/commit/67cfd0571e1e3b3b3228c1e0988e0aa3cd914b0a):

~~~editorconfig
root = true

[*]
charset = utf-8
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = false
indent_style = tab
indent_size = 4
~~~

As stated above, **this is a hack**. [I've been nagging GitHub for years](https://github.community/t/please-add-configurable-default-tab-size-already/121769/2) to add a repository setting for changing the default tab-width, but despite having the time to redesign the entire site (twice), they still haven't bothered implementing it. 🙄

**Comment by Alhadis** [comment]
If you have a moment, I recommend voicing your support at the [aforementioned thread](https://github.community/t/please-add-configurable-default-tab-size-already/121769/2) and/or [submitting a ticket](https://support.github.com/) to site support. The only way to get GitHub to listen to their users is when enough of them complain… 😖

(If I sound ticked-off, it's because I am. There's literally no reason they *shouldn't* add this, and it's not even complicated…)

**Comment by RealRaven2000** [comment]
> 
> 
> Does the current script not work when run as a bookmarklet in Firefox? thinking
> 

I don't know, I am not using Firefox this much. Just for some browsing task for which Waterfox Classic is too outdated. As long as there is no [easy way to get to the passwords](http://quickpasswords.quickfolders.org/) it has only limited usefulness to me as a browser.

> To change the tab-size GitHub uses when displaying source code, there's an unintuitive (and undocumented) hack you can try that involves adding an [`.editorconfig` file](https://editorconfig.org/) with an `indent_size` field. [For example](https://github.com/Alhadis/Utils/commit/67cfd0571e1e3b3b3228c1e0988e0aa3cd914b0a):

I can try it out, I need it mainly for diffs and viewing code (not so much for editing on the site, I really don't do that)

**Comment by Alhadis** [comment]
> I need it mainly for diffs and viewing code

I don't believe this trick extends to diffs, or to code-blocks in rendered markdown (including comments in PR/issue discussions). I'd report these oversights to GitHub, but I doubt they'd give a damn about fixing it…

> I don't know, I am not using Firefox this much

My bad, I should have said _\*Mozilla_. A bad habit of mine is referring to Gecko or Mozilla as "Firefox" (even though I know damn well that the latter refers to a browser, not a rendering engine or some other technology used in Firefox… 😅)

**Comment by RealRaven2000** [comment]
> 
> 
> > I don't know, I am not using Firefox this much
> 
> My bad, I should have said _*Mozilla_. A bad habit of mine is referring to Gecko or Mozilla as "Firefox" (even though I know damn well that the latter refers to a browser, not a rendering engine or some other technology used in Firefox… sweat_smile)

understandable mistake - I used PaleMoon for a while (for the same reason, mainly to use my own Add-ons QuickPasswords, MenuOnTop and ZombieKeys) and then found Waterfox, which is the most up-to-date app supporting legacy Add-ons. Thunderbird is a different kettle of fish, I have invested quite a bit of money so far and we are working on converting my commercial Add-ons to the interim status (mail experiments), with a view of hopefully getting them to work as web extensions next year (for this, I invested 3k so far).

Estimated effort for making web extensions: probably around 8,000$
So I need to seriously raise my user count to make it worth my while.

**Comment by Alhadis** [comment]
Well, good luck!

> Estimated effort for making web extensions: probably around 8,000$

To me, any figure longer than 3 digits sounds exorbitant. 😂 Then again, I'm broke as fuck.

> and a lot of it is "mixed" so the indentation looks useless with other tab sizes.

BTW, before I forget, a known limitation of this script is that it (currently) doesn't handle mixed or hanging indents. I intend to fix this eventually — it's more complicated than simple search-and-replace, as it needs to consider physical coordinates of HTML tags (to distinguish a semantic indent from a hanging indent).

This problem is compounded by the fact there's no easy way to retrieve the bounding-box of a purely textual node: i.e., retrieving the width of the word `foo` inside the following HTML tag:

~~~html
<span>abc foo xyz</span>
~~~

A workaround is to inject HTML tags around `foo`, but web-apps can monitor such changes and respond in ways that inhibit further manipulation. It can also report inaccurate metrics if it's targeted by a structural CSS selector:

~~~css
/* In the above example, this would appear to insert a new line after `abc `: */
span > * {
	display: block;
}
~~~

&lt;/FMTEYEWTK&gt;
